From a82022453a0b2b5b98074a7e06e0424330d36af0 Mon Sep 17 00:00:00 2001
From: John Jacques <john.jacques@intel.com>
Date: Mon, 17 Jul 2017 14:16:12 -0500
Subject: [PATCH 238/278] drivers/pci: Axxia Hardware Work Around

The Axxia pci controller has the wrong class code.  This
commit updates the code each time the controller gets
reset.

Signed-off-by: John Jacques <john.jacques@intel.com>
---
 drivers/pci/host/pcie-axxia.c | 34 +++++++++++++++++-----------------
 1 file changed, 17 insertions(+), 17 deletions(-)

diff --git a/drivers/pci/host/pcie-axxia.c b/drivers/pci/host/pcie-axxia.c
index 32fc728..cc7891d 100644
--- a/drivers/pci/host/pcie-axxia.c
+++ b/drivers/pci/host/pcie-axxia.c
@@ -893,6 +893,20 @@ int axxia_pcie_link_up(struct pcie_port *pp)
 	return rc;
 }
 
+static void pci_axxia_program_rc_class(struct pcie_port *pp)
+{
+	u32 dbi_ro_wr_en;
+	/* program correct class for RC */
+	axxia_pcie_readl_rc(pp, 0x8bc, &dbi_ro_wr_en);
+	/* DBI_RO_WR_EN */
+	if (!(dbi_ro_wr_en & 0x1))
+		axxia_pcie_writel_rc(pp, (dbi_ro_wr_en | 0x1), 0x8bc);
+	axxia_pcie_wr_own_conf(pp, PCI_CLASS_DEVICE, 2, PCI_CLASS_BRIDGE_PCI);
+	/* DBI_RO_WR_EN */
+	if (!(dbi_ro_wr_en & 0x1))
+		axxia_pcie_writel_rc(pp, dbi_ro_wr_en, 0x8bc);
+}
+
 void axxia_pcie_setup_rc(struct pcie_port *pp)
 {
 	u32 val;
@@ -906,6 +920,9 @@ void axxia_pcie_setup_rc(struct pcie_port *pp)
 		return;
 
 	for (;;) {
+		/* program correct class for RC */
+		pci_axxia_program_rc_class(pp);
+
 		/*
 		  To work around a hardware problem, set
 		  PCIE_LINK_WIDTH_SPEED_CONTROL to 1 lane in all cases.
@@ -1018,20 +1035,6 @@ static int axxia_pcie_establish_link(struct pcie_port *pp)
 	return 0;
 }
 
-static void pci_axxia_program_rc_class(struct pcie_port *pp)
-{
-	u32 dbi_ro_wr_en;
-	/* program correct class for RC */
-	axxia_pcie_readl_rc(pp, 0x8bc, &dbi_ro_wr_en);
-	/* DBI_RO_WR_EN */
-	if (!(dbi_ro_wr_en & 0x1))
-		axxia_pcie_writel_rc(pp, (dbi_ro_wr_en | 0x1), 0x8bc);
-	axxia_pcie_wr_own_conf(pp, PCI_CLASS_DEVICE, 2, PCI_CLASS_BRIDGE_PCI);
-	/* DBI_RO_WR_EN */
-	if (!(dbi_ro_wr_en & 0x1))
-		axxia_pcie_writel_rc(pp, dbi_ro_wr_en, 0x8bc);
-}
-
 static irqreturn_t axxia_pcie_irq_handler(int irq, void *arg)
 {
 	struct pcie_port *pp = arg;
@@ -1298,9 +1301,6 @@ int axxia_pcie_host_init(struct pcie_port *pp)
 		return -EINVAL;
 	}
 
-	/* program correct class for RC */
-	pci_axxia_program_rc_class(pp);
-
 	if (axxia_pcie_establish_link(pp))
 		dev_warn(pp->dev, "axxia_pcie_establish_link failed\n");
 
-- 
1.8.5.2.233.g932f7e4

